qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,316 @@
+        self._update_config(
+            self.ensemble_config_file,
+            self.decoupled_pattern,
+            self.decoupled_config,
+        )
+
+    def enable_decoupled_composing_model(self):
+        self._update_config(
+            self.composing_config_file, self.decoupled_pattern, self.decoupled_config
+        )
+
+
+class InferenceHandler:
+    def __init__(self):
+        self.config_manager = ModelConfigManager()
+        self.input_tensors = self._get_input_tensors()
+        self.triton_client = grpcclient.InferenceServerClient(
+            "localhost:8001", verbose=True
+        )
+
+    def _get_input_tensors(self):
+        input0_data = np.ones((1, 16), dtype=np.float32)
+        input1_data = np.ones((1, 16), dtype=np.float32)
+        input_tensors = [
+            grpcclient.InferInput(
+                "INPUT0", input0_data.shape, np_to_triton_dtype(input0_data.dtype)
+            ),
+            grpcclient.InferInput(
+                "INPUT1", input1_data.shape, np_to_triton_dtype(input0_data.dtype)
+            ),
+        ]
+        input_tensors[0].set_data_from_numpy(input0_data)
+        input_tensors[1].set_data_from_numpy(input1_data)
+        return input_tensors
+
+    def get_inference_statistics(self, model):
+        stats = self.triton_client.get_inference_statistics(
+            model_name=model, as_json=True
+        )
+        return stats["model_stats"][1]["inference_stats"]
+
+    def infer(self):
+        model = self.config_manager.get_ensemble_model()
+        output = self.triton_client.infer(model_name=model, inputs=self.input_tensors)
+        output0 = output.as_numpy("OUTPUT0")
+        output1 = output.as_numpy("OUTPUT1")
+        outputs = [output0, output1]
+        return outputs
+
+
+class InferenceValidator:
+    def __init__(self):
+        self.inference_handler = InferenceHandler()
+
+    def _check_valid_output(self, output):
+        if output is None:
+            self.assertTrue(False, "unexpected error in inference")
+
+    def _check_valid_output_inference(self, inference_outputs, cached_outputs):
+        if not np.array_equal(inference_outputs, cached_outputs):
+            self.assertTrue(False, "mismtached outputs")
+
+    def _check_zero_stats_baseline(self, model_inference_stats):
+        if "count" in model_inference_stats["success"]:
+            self.assertTrue(False, "unexpected non-zero inference statistics")
+
+    def _check_valid_stats(self, model_inference_stats):
+        if "count" not in model_inference_stats["success"]:
+            self.assertTrue(False, "unexpected error while retrieving statistics")
+
+    def _check_single_cache_miss_success_inference(self, model_inference_stats):
+        if (
+            "count" not in model_inference_stats["cache_miss"]
+            and "count" not in model_inference_stats["cache_hit"]
+        ):
+            self.assertTrue(False, "unexpected error with response cache")
```

Comment:
I'm not sure I fully understand what is happening here. 

If we want to make sure that both `"count" in model_inference_stats["cache_miss"]` and `"count"  in model_inference_stats["cache_hit"]` are True, then I would suggest:

```
expr = "count" in model_inference_stats["cache_miss"] and "count"  in model_inference_stats["cache_hit"]
assertTrue(expr, msg="unexpected error with response cache")
```
reference: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertTrue 

`self.assertTrue(False, ...)` is a bit confusing, i.e. I know that False is not True, what we're asserting here?

Reply:
I followed this https://github.com/triton-inference-server/server/blob/main/qa/L0_simple_ensemble/ensemble_test.py
when I first started writing testcases. 
I will make changes as suggested.

Reply:
No problem, that test was added in 2019, not sure what state python unittests library was in, but now it has a variety of methods, which worth exploring and using, since they provide better messages and are tailored to what you are trying to check